File: src/Muflone/Core/Entity.cs
﻿using System;

namespace Muflone.Core;

public abstract class Entity : IEquatable<Entity>
{
	public readonly IDomainId Id = default!;

	protected Entity()
	{
	}

	protected Entity(IDomainId id)
	{
		Id = id;
	}

	public override int GetHashCode()
	{
		return Id.Value.GetHashCode();
	}

	public override bool Equals(object? obj)
	{
		return Equals(obj as Entity);
	}

	public bool Equals(Entity? other)
	{
		return (null != other) && (GetType() == other.GetType()) && (other.Id.Value == Id.Value);
	}

	public static bool operator ==(Entity? entity1, Entity? entity2)
	{
		if (entity1 is null && entity2 is null)
			return true;

		if (entity1 is null || entity2 is null)
			return false;

		return (entity1.GetType() == entity2.GetType()) && (entity1.Id.Value == entity2.Id.Value);
	}

	public static bool operator !=(Entity? entity1, Entity? entity2)
	{
		return !(entity1 == entity2);
	}
}

File: php/src/Muflone/Core/Entity.php
<?php

declare(strict_types=1);

namespace Muflone\Core;

abstract class Entity
{
    protected readonly IDomainId $id;

    public function __construct(?IDomainId $id = null)
    {
        if ($id !== null) {
            $this->id = $id;
        }
    }

    public function getId(): IDomainId
    {
        return $this->id;
    }

    public function equals(?self $other): bool
    {
        if ($other === null) {
            return false;
        }

        return get_class($this) === get_class($other)
            && $other->id->getValue() === $this->id->getValue();
    }

    public function __hash(): string
    {
        return hash('sha256', $this->id->getValue());
    }
}
